mesonbuild/modules/gnome.py

```diff
@@ -787,12 +790,28 @@ def postconf_hook(self, b: build.Build) -> None:
         if self.devenv is not None:
             b.devenv.append(self.devenv)
 
-    def _get_gir_dep(self, state: 'ModuleState') -> T.Tuple[Dependency, T.Union[Executable, 'ExternalProgram', 'OverrideProgram'],
-                                                            T.Union[Executable, 'ExternalProgram', 'OverrideProgram']]:
+    def _supports_girepository2(self, state: 'ModuleState') -> bool:
+        glib_version = self._get_native_glib_version(state)
+
+        return mesonlib.version_compare(glib_version, '>= 2.79.2')
+
+    def _get_gir_dep(self, state: 'ModuleState', use_gi_repository_compile: bool
+                     ) -> T.Tuple[Dependency, T.Union[Executable, 'ExternalProgram', 'OverrideProgram'],
+                                  T.Union[Executable, 'ExternalProgram', 'OverrideProgram']]:
+        supports_girepository_2 = self._supports_girepository2(state)
+        if supports_girepository_2 and not self.girepository_dep and use_gi_repository_compile:
+            self.girepository_dep = state.dependency('girepository-2.0')
+
         if not self.gir_dep:
             self.gir_dep = state.dependency('gobject-introspection-1.0')
             self.giscanner = self._find_tool(state, 'g-ir-scanner')
-            self.gicompiler = self._find_tool(state, 'g-ir-compiler')
+
+        if not self.gicompiler:
+            if use_gi_repository_compile and supports_girepository_2:
+                self.gicompiler = self._find_tool(state, 'gi-compile-repository')
+            else:
+                self.gicompiler = self._find_tool(state, 'g-ir-compiler')
+
         return self.gir_dep, self.giscanner, self.gicompiler
```

Comment:
Implementation-wise, this is very problematic. The current design requires people to specify which compiler they want via a meson.build DSL keyword argument, but once passed for the first time it will cache the value globally.

This means for example that if you have a project which knows nothing about this new feature -- it supports meson 0.40.0, or more charitably, supports meson 1.6.0 -- it will configure to build a gir using g-ir-compiler, and then it invokes a subproject that has a minimum `meson_version: '>=1.7.0'` and makes heavy use of GIRepository 2.0 features and is simply incompatible with GIRepository 1.0... and fails to compile, even though it passes your `use_gi_repository_compile: true` kwarg, because meson simply detects that `self.gicompiler` is g-ir-compiler.

This is one of the reasons I was hoping it would be practical to simply use the newer tool by default. :)